src/lib.rs

```diff
@@ -447,11 +455,98 @@ impl Database {
+            },
         }
     }
-}
 
+    /// Makes the font data that backs the specified face id shared so that the application can
+    /// hold a reference to it.
+    ///
+    /// # Safety
+    ///
+    /// If the face originates from a file from disk, then the file is mapped from disk. This is unsafe as
+    /// another process may make changes to the file on disk, which may become visible in this process'
+    /// mapping and possibly cause crashes.
+    ///
+    /// If the underlying font provides multiple faces, then all faces are updated to participate in
+    /// the data sharing. If the face was previously marked for data sharing, then this function will
+    /// return a clone of the existing reference.
+    pub unsafe fn make_shared_face_data(&mut self, id: ID) -> Option<(Arc<dyn AsRef<[u8]> + Send + Sync>, u32)> {
+        let face_info = self.faces.iter().find(|item| item.id == id)?;
+        let face_index = face_info.index;
+
+        let old_source = face_info.source.clone();
+
+        let (new_source, erased_shared_data) = match old_source.as_ref() {
+            Source::Binary(data) => {
+                let shared_data = Arc::new(data.clone());
+                (Source::SharedBinary(shared_data.clone()), shared_data as Arc<dyn AsRef<[u8]> + Send + Sync>)
+            },
+            #[cfg(all(feature = "fs", not(feature = "memmap")))]
+            Source::File(ref path) => {
+                let data = std::fs::read(path).ok()?;
+                (Some(path.clone()), Arc::new(data) as Arc<dyn AsRef<[u8]> + Send + Sync>)
+            }
+            #[cfg(all(feature = "fs", feature = "memmap"))]
+            Source::File(ref path) => {
+                let file = std::fs::File::open(path).ok()?;
+                let shared_data = Arc::new(memmap2::MmapOptions::new().map(&file).ok()?) as Arc<dyn AsRef<[u8]> + Send + Sync>;
+                (Source::SharedFile(path.clone(), shared_data.clone()), shared_data)
+            }
+            #[cfg(all(feature = "fs", feature = "memmap"))]
+            Source::SharedFile(_, data) => {
+                return Some((data.clone(), face_index));
+            }
+            Source::SharedBinary(data) => {
+                return Some((data.clone(), face_index));
+            }
+        };
+
+        let shared_source = Arc::new(new_source);
+
+        self.faces.iter_mut().for_each(|face| {
+            if Arc::ptr_eq(&face.source, &old_source) {
+                face.source = shared_source.clone();
+            }
+        });
+
+        Some((erased_shared_data, face_index))
+    }
+
+    /// Transfers ownership of shared font data back to the font database. This is the reverse operation
+    /// of [`Self::make_shared_face_data`]. If the font data belonging to the specified face is mapped
+    /// from a file on disk, then that mapping is closed and the data becomes private to the process again.
+    pub fn make_face_data_unshared(&mut self, id: ID) {
```

Comment:
This function is not marked as unsafe, but for "symmetry" it could be. I don't have a strong opinion on it - happy to adjust either way.

Reply:
I guess it's fine.